src/lib.rs

```diff
@@ -116,6 +130,99 @@ where
     }
 }
 
+struct FullPermutations<T, const N: usize> {
+    a: [T; N],
+    c: [usize; N],
+    i: usize,
+    first: bool,
+}
+
+impl<T, const N: usize> FullPermutations<T, N> {
+    fn new(a: [T; N]) -> Self {
+        Self {
+            a,
+            c: [0; N],
+            i: 0,
```

Comment:
Probably something for a follow-up PR; but it'd be good to have more descriptive names for these variables. If we want to add this to the stdlib we should expect this code will be maintained for many years to come by people other than us.

Reply:
I agree, for now I kept the names the same as in the pseudocode, but we should not depend on that and we should make it clear how it works.